core/utils/codel.h

```diff
@@ -0,0 +1,331 @@
+
+  /**
+   * Gets the next object from the queue and determines based on current state,
+   * whether to mark in the wrapper that the object should be dropped.
+   * Takes an integer pointer to set on error.
+   */
+  Wrapper* RingDequeue(int* err);
+
+  /**
+   * Called while Codel is in drop state to determine whether to drop the current
+   * entries and dequeue the next entry. Will continue to drop entries until the next
+   * time till drop is greater than the current time. Takes a pointer to the wrapper
+   * pointer where the current entry is located and an integer pointer to set on 
+   * error.
+   */
+  Wrapper* DropDequeue(Wrapper* w, int* err);
+
+  // Returns the current time in microseconds.
+  uint64_t MicroSecondTime();
+
+  uint64_t delay_target_;  // the delay that codel will adjust for
+  uint64_t window_;        // minimum time before changing state
+  // the time at which codel will change state to above target(0 if below)
+  uint64_t time_above_target_;
+  uint64_t time_till_drop_;  // the next time codel will drop
+  // the number of packets dropped while delay has been above target
+  uint32_t drop_count_;
+  uint8_t dropping_;       // whether in dropping state(above target for window)
+  Queue<Wrapper>* queue_;  // queue
+  void (*drop_func_)(T*);  // the function to call to drop a value
+};
+
+template <template <typename type> class Q, typename T>
+Codel<Q, T>::Codel(void (*drop_func)(T*), Q<Wrapper>* init_queue,
+                   uint64_t target, uint64_t window)
+    : delay_target_(target),
+      window_(window),
+      time_above_target_(0),
+      time_till_drop_(MicroSecondTime() + window),
+      drop_count_(0),
+      dropping_(0),
+      queue_(init_queue),
+      drop_func_(drop_func) {}
+
+template <template <typename type> class Q, typename T>
+Codel<Q, T>::~Codel() {
+  Wrapper* w;
+  int err = 0;
+  while ((w = queue_->Pop(&err))) {
+    Drop(w);
+  }
+}
+
+template <template <typename type> class Q, typename T>
+bool Codel<Q, T>::Full() {
+  return queue_->Full();
+}
+
+template <template <typename type> class Q, typename T>
+bool Codel<Q, T>::Empty() {
+  return queue_->Empty();
+}
+
+template <template <typename type> class Q, typename T>
+size_t Codel<Q, T>::Size() {
+  return queue_->Size();
+}
+
+template <template <typename type> class Q, typename T>
+size_t Codel<Q, T>::Capacity() {
+  return queue_->Capacity();
+}
+
+template <template <typename type> class Q, typename T>
+uint64_t Codel<Q, T>::MicroSecondTime() {
+  struct timeval tv;
+  gettimeofday(&tv, nullptr);
+  return tv.tv_sec * 1e6 + tv.tv_usec;
+}
+
+template <template <typename type> class Q, typename T>
+void Codel<Q, T>::Drop(Wrapper* w) {
+  if (w != nullptr) {
+    if (w->ptr != nullptr) {
+      drop_func_(w->ptr);
+    }
+    delete w;
+  }
+}
+
+template <template <typename type> class Q, typename T>
+uint64_t Codel<Q, T>::NextDrop(uint64_t cur_time) {
+  return cur_time + window_ * pow(drop_count_, -.5);
```

Comment:
we decided to leave pow as there are not any readily available built-ins that work for this purpose. however, this might be changed to more closely resemble the kernel implementation if it becomes a bottleneck.